coderd/userauth.go

```diff
@@ -64,6 +67,49 @@ func (api *API) userOAuth2Github(rw http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	// The default if no teams are specified is to allow all.
+	if len(api.GithubOAuth2Config.AllowTeams) > 0 {
+		teams, err := api.GithubOAuth2Config.ListTeams(r.Context(), oauthClient, *selectedMembership.Organization.Login)
+		if err != nil {
+			httpapi.Write(rw, http.StatusInternalServerError, httpapi.Response{
+				Message: "Failed to fetch teams from GitHub.",
+				Detail:  err.Error(),
+			})
+			return
+		}
+
+		var allowedTeam *github.Team
+		for _, team := range teams {
+			for _, organizationAndTeam := range api.GithubOAuth2Config.AllowTeams {
+				parts := strings.SplitN(organizationAndTeam, "/", 2)
+				if len(parts) != 2 {
+					httpapi.Write(rw, http.StatusInternalServerError, httpapi.Response{
+						Message: "Team allowlist isn't formatted correctly.",
+						Detail:  fmt.Sprintf("Got %s, wanted <organization>/<team>", organizationAndTeam),
+					})
```

Comment:
I mentioned we should check this earlier, but if we are ok with delaying the parse until here, we should also api.Logger.Error log this failure.